### PR TITLE
feat(rh-table): add rh-table

### LINKED ISSUE #431: [feat] <rh-table>
### Description

This component must accommodate the need for a standardized table data style for all digital experiences. This component exists in the Patternfly Design System, but the purpose was to support large data sets. This component will support large data sets and multiline content types.

### Acceptance Criteria

- [x] Design done
- [ ] Development done

### Image

table

### Link to design doc

https://xd.adobe.com/view/a598a299-20af-4a97-b37e-cbc04931e112-f0cb/screen/0ac8b4cf-f544-4cfb-8a2e-7600d5db09ee/

### Other resources

https://docs.google.com/document/d/1BHsSVS99DMmTUqEKFTRz7pPLB2ZhcNF1fzOOhCIfqPQ/edit#heading=h.nui944fdk774

### PR DESCRIPTION
**Pull request by eyevana**
## What I did

1. Created a new `rh-table` element based on the following [design spec](https://www.figma.com/file/55pIOeQu2ry4oaGvhrkLjM/AsciiDoc-Elements?type=design&node-id=251-284&mode=design&t=n55MkCb4JlYbMuOk-0).

Closes #431.

### CODE REVIEW

**Comment by eyevana** [comment]
@nikkimk @hellogreg in the [table wires](https://www.figma.com/file/55pIOeQu2ry4oaGvhrkLjM/AsciiDoc-Elements?type=design&node-id=15-536&mode=design&t=05F1w42Ymp8kREWR-0), they refer to a table header (i.e. caption) and a table caption (what I'm thinking is more of a summary/disclaimer). After reading through these [docs](https://www.w3.org/WAI/tutorials/tables/caption-summary/), I decided to go with option two of linking the table with the summary via a named slot with `aria-describedby`. What do you think of this approach over the other options? Option one sounds like it has better support, but I'm not sure how I could get the summary to show up under the table for sighted users.

**Comment by hellogreg** [comment]
In checking the WAI example, `aria-describedby` worked fine for me with the main browser / screen reader combos. If you got that route, just make sure you've got it pointing to a valid `id` with no cross-root issues.

Also, I don't know where you are on finalizing the sortable column header buttons, but it looks like they currently have text like "(sort in descending order)"--and don't include the header text. The WAI Sortable Table example uses the header text in the button and recommends the following:

> To help screen reader users understand the purpose of the buttons in the column headers, an off-screen description of the sort functionality of the buttons is appended to the caption text. The description is added to the caption instead of to each button to prevent repetitious verbosity that could interfere with understanding of the column titles.

In other examples by reputable sources, [Adrian Roselli's sortable tables](https://adrianroselli.com/2021/04/sortable-table-columns.html) just read the header text, but [Heydon Pickering's tables](https://heydon.github.io/react-sortable-table-demo/) _do_ make mention of the both the header name and whether the columns are sortable (and announce the direction).

Nikki may have ideas on this, too.

**Comment by brianferry** [comment]
From looking at the `<col/>` elements it looks like that's a good way to go.  There are [CSS tricks ](https://css-tricks.com/simple-css-row-column-highlighting/) to get around having to include the `<col/>` elements but they seem more like hacks than standards.  This will probably just be a pattern configuration imo though so we might be able to abstract that out a bit.

File: elements/rh-table/rh-table.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+import { RequestSortEvent, RhSortButton } from './rh-sort-button.js';
+export * from './rh-sort-button.js';
+
+/**
+ * Displays tabular information in an easy and digestable manner.
+ *
+ * @summary
+ *
+ * @slot            - an HTML table
+ * @slot    summary - a brief description of the data
```
Comment (zeroedin) [inline elements/rh-table/rh-table.ts:16]: Add a @summary

As well @csspart and @cssprop usage here as well

File: elements/rh-table/rh-table.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+import { customElement } from 'lit/decorators/custom-element.js';
+
+import { Logger } from '@patternfly/pfe-core/controllers/logger.js';
+
+import styles from './rh-table.css';
+import { RequestSortEvent, RhSortButton } from './rh-sort-button.js';
+export * from './rh-sort-button.js';
+
+/**
+ * Displays tabular information in an easy and digestable manner.
```
Comment (zeroedin) [inline elements/rh-table/rh-table.ts:11]: spelling
```Suggestion
 * Displays tabular information in an easy and digestible manner.
```

File: elements/rh-table/rh-table.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
+import { LitElement, html } from 'lit';
+import { customElement } from 'lit/decorators/custom-element.js';
+
+import { Logger } from '@patternfly/pfe-core/controllers/logger.js';
+
+import styles from './rh-table.css';
+import { RequestSortEvent, RhSortButton } from './rh-sort-button.js';
+export * from './rh-sort-button.js';
```
Comment (zeroedin) [inline elements/rh-table/rh-table.ts:8]: Not sure this export needs to be here.   IIRC that would already be exported at the package level from `"exports": { ./*": "./elements/*"}` in package.json

File: elements/rh-table/rh-table.ts [hunk 4]
```diff
@@ -0,0 +1,168 @@
+import { LitElement, html } from 'lit';
+import { customElement } from 'lit/decorators/custom-element.js';
+
+import { Logger } from '@patternfly/pfe-core/controllers/logger.js';
+
+import styles from './rh-table.css';
+import { RequestSortEvent, RhSortButton } from './rh-sort-button.js';
```
Comment (zeroedin) [inline elements/rh-table/rh-table.ts:7]:
```Suggestion
import { type RequestSortEvent, RhSortButton } from './rh-sort-button.js';
```

File: elements/rh-table/rh-table.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+  }
+
+  #logger = new Logger(this);
+
+  render() {
+    return html`
+      <div id="container"
+           @pointerleave=${this.#onPointerleave}
+           @pointerover=${this.#onPointerover}>
+        <slot @request-sort="${this.#onRequestSort}" @slotchange=${this.#onSlotChange}></slot>
```
Comment (zeroedin) [inline elements/rh-table/rh-table.ts:41]: Nice usage of a declarative listener.  Didn't even think about using them for custom events, just things like `@click`

File: elements/rh-table/rh-table.ts [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        col.classList.add('active');
+      } else {
+        col.classList.remove('active');
+      }
+    });
+  }
+
+  #onSlotChange() {
+    if (this.#table) {
+      this.#table.setAttribute('aria-describedby', 'summary');
```
Comment (zeroedin) [inline elements/rh-table/rh-table.ts:89]: Would/Should this only happen if the summary slot had content?  @nikkimk

File: elements/rh-table/rh-sort-button.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+export class RequestSortEvent extends Event {
+  constructor(
+      public direction: 'asc' | 'desc',
+  ) {
+    super('request-sort', {
+      bubbles: true,
+      cancelable: true,
+    });
+  }
+}
```
Comment (zeroedin) [inline elements/rh-table/rh-sort-button.ts:18]: I think you could use `ComposedEvent` here

`import { ComposedEvent } from '@patternfly/pfe-core';`

```Suggestion
export class RequestSortEvent extends ComposedEvent {
  constructor(
      public direction: 'asc' | 'desc',
  ) {
    super('request-sort', { cancelable: true });
  }
}
```

**Review by zeroedin — CHANGES_REQUESTED** [review]
don't forget to add a changeset and update screenshot.png

File: elements/rh-table/rh-table.ts (hunk 6 above)
Comment (eyevana) [reply elements/rh-table/rh-table.ts:94]: Good catch. I'm checking for a slotted summary as well now.

File: elements/rh-table/rh-table.ts (hunk 1 above)
Comment (eyevana) [reply elements/rh-table/rh-table.ts]: Waiting for @summary from Corey

File: elements/rh-table/rh-table.ts (hunk 3 above)
Comment (eyevana) [reply elements/rh-table/rh-table.ts]: Done! Thx ")

File: elements/rh-table/rh-table.ts (hunk 1 above)
Comment (eyevana) [reply elements/rh-table/rh-table.ts]: Added ✔️

**Review by coreyvickery — APPROVED** [review]
Looks good.

File: .changeset/healthy-emus-destroy.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+```html
+<rh-table>
+  <table>
+    <caption>
+      Concerts
+    </caption>
+    <colgroup>
+      <col />
+      <col />
+      <col />
```
Comment (bennypowers) [inline .changeset/healthy-emus-destroy.md:18]:
```suggestion
      <col>
      <col>
      <col>
```
although mdn's `<col>` docs include them, [void elements shouldn't have a slash](https://html.spec.whatwg.org/multipage/syntax.html#start-tags)

File: elements/rh-table/README.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
+# Table
+
+A table is a container for displaying information. It allows a user to scan, examine, and compare large amounts of data.
```
Comment (bennypowers) [inline elements/rh-table/README.md:3]: each of our progressive enhancements (sorting, etc) should be listed here

File: elements/rh-table/README.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+```html
+<rh-table>
+  <table>
+    <caption>
+      Concerts
+    </caption>
+    <colgroup>
+      <col />
+      <col />
+      <col />
```
Comment (bennypowers) [inline elements/rh-table/README.md:16]:
```suggestion
      <col>
      <col>
      <col>
```

File: elements/rh-table/demo/rh-table.html
Comment (bennypowers) [inline elements/rh-table/demo/rh-table.html:1]: break these up into separate files with the snippet from the changeset as the canonical demo

File: elements/rh-table/demo/rh-table.js
Comment (bennypowers) [inline elements/rh-table/demo/rh-table.js:1]: prefer to inline this into demos, to improve reading comprehension

File: elements/rh-table/docs/00-overview.md [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+## Sample element
+
+<rh-table>
+  <table>
+    <colgroup>
+      <col/>
+      <col/>
+      <col/>
+      <col/>
```
Comment (bennypowers) [inline elements/rh-table/docs/00-overview.md:17]:
```suggestion
      <col>
      <col>
      <col>
      <col>
```

File: elements/rh-table/docs/20-guidelines.md [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+## Character count
+
+In general, header labels should be as short as possible. However, if columns have more width, more words can be added.
+
+<rh-table>
+  <table>
+    <colgroup>
+        <col style="width: 25%" />
+        <col />
```
Comment (bennypowers) [inline elements/rh-table/docs/20-guidelines.md:39]:
```suggestion
        <col style="width: 25%">
        <col>
```

File: elements/rh-table/docs/30-code.md [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+```html
+<rh-table>
+  <table>
+    <caption>
+      Concerts
+    </caption>
+    <colgroup>
+      <col />
+      <col />
+      <col />
```
Comment (bennypowers) [inline elements/rh-table/docs/30-code.md:33]:
```suggestion
      <col>
      <col>
      <col>
```

File: elements/rh-table/rh-sort-button.ts [hunk 14]
```diff
@@ -0,0 +1,67 @@
+import { LitElement, html, svg } from 'lit';
+import { customElement } from 'lit/decorators/custom-element.js';
+import { property } from 'lit/decorators/property.js';
+import styles from './rh-sort-button.css';
+import { ComposedEvent } from '@patternfly/pfe-core';
+
+const DIRECTIONS = { asc: 'desc', desc: 'asc' } as const;
```
Comment (bennypowers) [inline elements/rh-table/rh-sort-button.ts:7]:
```suggestion
const DIRECTION_OPPOSITES = { asc: 'desc', desc: 'asc' } as const;
```

File: elements/rh-table/rh-sort-button.css [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+  position: absolute;
+  inset: 0;
+  cursor: pointer;
+}
+
+#sort-button #sort-indicator {
+  color: currentcolor;
+}
+
+.visually-hidden {
```
Comment (bennypowers) [inline elements/rh-table/rh-sort-button.css:17]: at this point maybe we should prepare a shared stylesheet with this. thoughts @zeroedin ?

File: elements/rh-table/rh-sort-button.ts [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+@customElement('rh-sort-button')
+export class RhSortButton extends LitElement {
+  static readonly styles = [styles];
+
+  @property({
+    reflect: true,
+    attribute: 'sort-direction',
+  }) sortDirection?: 'asc' | 'desc';
+
+  @property() column?: string;
```
Comment (bennypowers) [inline elements/rh-table/rh-sort-button.ts:36]: jsdoc plz

File: elements/rh-table/rh-table.ts [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+
+  #logger = new Logger(this);
+
+  render() {
+    return html`
+      <div id="container"
+           part="container"
+           @pointerleave=${this.#onPointerleave}
+           @pointerover=${this.#onPointerover}>
+        <slot @request-sort="${this.#onRequestSort}" @slotchange=${this.#onSlotChange}></slot>
```
Comment (bennypowers) [inline elements/rh-table/rh-table.ts:46]: consider removing the wrapping div and setting a display property on the slot

File: elements/rh-table/rh-table.ts [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+  #onPointerover(event: PointerEvent) {
+    if (!this.#cols) {
+      return;
+    }
+
+    event.preventDefault();
+
+    let { target } = event;
+
+    if (!(target instanceof Element)) {
```
Comment (bennypowers) [inline elements/rh-table/rh-table.ts:69]: should the  prevent default happen after all the short circuits?

File: elements/rh-table/rh-table.ts [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+      }
+    }
+
+    this.#cols.forEach((col, index) => {
+      const { cellIndex } = target as HTMLTableCellElement;
+      if (index === cellIndex) {
+        col.classList.add('active');
+      } else {
+        col.classList.remove('active');
+      }
```
Comment (bennypowers) [inline elements/rh-table/rh-table.ts:88]:
```suggestion
      col.classList.toggle('active', index === cellIndex);
```

Comment (bennypowers) [reply elements/rh-table/rh-table.ts:88]: https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/toggle

File: elements/rh-table/rh-table.ts [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+    });
+  }
+
+  #onSlotChange() {
+    if (this.#table && this.#summary) {
```
Comment (bennypowers) [inline elements/rh-table/rh-table.ts:93]: ensure this happens on connected as well

File: elements/rh-table/rh-table.ts [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+      this.#table.setAttribute('aria-describedby', 'summary');
+    }
+  }
+
+  #onRequestSort(event: Event) {
+    if (event instanceof RequestSortEvent) {
+      for (const button of this.querySelectorAll<RhSortButton>('rh-sort-button')) {
```
Comment (bennypowers) [inline elements/rh-table/rh-table.ts:100]: not sure we need the type assertion on qsa here. as long as the selector is just the tagname, typescript should pick it up

File: elements/rh-table/rh-table.ts [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+
+  // @todo: should we move the remaining methods into a controller to share with pf-table?
+  #performSort(button: RhSortButton, direction: 'asc' | 'desc') {
+    const header = button.closest('th');
+    const children = header?.parentElement?.children;
+    if (children) {
+      const columnIndexToSort = [...children].indexOf(header);
+
+      if (!this.#rows) {
+        this.#logger.warn('No rows found');
```
Comment (bennypowers) [inline elements/rh-table/rh-table.ts:124]:
```suggestion
        this.#logger.warn('Could not perform sort: no rows found');
```

File: elements/rh-table/rh-table.ts [hunk 23]
```diff
@@ -0,0 +1,173 @@
...
+          if (this.#rows[index] !== node) {
+            const position: InsertPosition =
+                direction === 'desc' ? 'afterend' : 'beforebegin';
+            target.insertAdjacentElement(position, node);
+          }
+        });
+    }
+  }
+
+  private static getNodeContentForSort(
```
Comment (bennypowers) [inline elements/rh-table/rh-table.ts:145]: member ordering. see wiki

**Review by bennypowers — CHANGES_REQUESTED** [review]
excellent work!

i didn't review the docs contents. Please recruit developers from studio dev and systems fed, etc (i.e. future users of rh-table) to make sure the docs are clear to them

**Comment by hellogreg** [comment]
Small heading text change for the demo page: "Not title, headers, and summary" to "No title, headers, or summary"

screencap of heading: Not title, headers, and summary

**Comment by eyevana** [comment]
@bennypowers thank you so much for the review <3 <3 <3 I've incorporated all of the requested changes. If anything new pops up, let me know and I'll fix asap :)

File: elements/rh-table/rh-sort-button.css (hunk 15 above)
Comment (zeroedin) [reply elements/rh-table/rh-sort-button.css:17]: I think its worthwhile to have a couple shared helpers, @nikkimk and I have discussed this before as well.

File: elements/rh-table/demo/different-layouts.html [hunk 24]
```diff
@@ -0,0 +1,131 @@
+<link rel="stylesheet" href="demo.css">
```
Comment (bennypowers) [inline elements/rh-table/demo/different-layouts.html:1]: inline this

File: elements/rh-table/demo/different-layouts.html
Comment (bennypowers) [inline elements/rh-table/demo/different-layouts.html:1]: i think each of these examples can be a separate file i.e. row-headers.html, column-headers.html, row-and-column-headers.html

File: elements/rh-table/demo/rh-table.html [hunk 25]
```diff
@@ -0,0 +1,49 @@
+<link rel="stylesheet" href="demo.css">
```
Comment (bennypowers) [inline elements/rh-table/demo/rh-table.html:1]: inline this.

File: elements/rh-table/demo/variants.html [hunk 26]
```diff
@@ -0,0 +1,319 @@
+<link rel="stylesheet" href="demo.css">
```
Comment (bennypowers) [inline elements/rh-table/demo/variants.html:1]: inline this

File: elements/rh-table/rh-table-lightdom.css [hunk 27]
```diff
@@ -0,0 +1,150 @@
...
+:is(rh-table) table {
+  min-width: 100%;
+  margin: 0 auto;
+  table-layout: fixed;
+  border: 0;
+  border-collapse: collapse;
+
+  --rh-table--row-border: var(--rh-border-width-sm, 1px) solid var(--rh-color-border-subtle-on-light, #c7c7c7);
+  --rh-table--row-background-color: rgb(var(--rh-color-gray-10-rgb, 224 224 224) / var(--rh-opacity-40, 40%));
+  --rh-table--column-background-color: rgb(var(--rh-color-blue-50-rgb, 231 241 250));
```
Comment (bennypowers) [inline elements/rh-table/rh-table-lightdom.css:11]:
```suggestion
  --rh-table-row-border: var(--rh-border-width-sm, 1px) solid var(--rh-color-border-subtle-on-light, #c7c7c7);
  --rh-table-row-background-color: rgb(var(--rh-color-gray-10-rgb, 224 224 224) / var(--rh-opacity-40, 40%));
  --rh-table-column-background-color: rgb(var(--rh-color-blue-50-rgb, 231 241 250));
```

File: elements/rh-table/rh-table.ts [hunk 28]
```diff
@@ -0,0 +1,175 @@
...
+    this.#init();
+  }
+
+  render() {
+    return html`
+        <slot @pointerleave=${this.#onPointerleave}
+              @pointerover=${this.#onPointerover}
+              @request-sort="${this.#onRequestSort}" 
+              @slotchange=${this.#onSlotChange}></slot>
+        <slot id="summary" name="summary"></slot>
```
Comment (bennypowers) [inline elements/rh-table/rh-table.ts:73]:
```suggestion
        <slot @pointerleave="${this.#onPointerleave}"
              @pointerover="${this.#onPointerover}"
              @request-sort="${this.#onRequestSort}" 
              @slotchange="${this.#onSlotChange}"></slot>
        <slot id="summary" name="summary"></slot>
```

help treesitter cope with life

**Review by bennypowers — CHANGES_REQUESTED** [review]
almost there!

File: elements/rh-table/rh-table-lightdom.css (hunk 27 above)
Comment (bennypowers) [reply elements/rh-table/rh-table-lightdom.css:11]: check all other instances, too

**Review by bennypowers — APPROVED** [review]

**Review by zeroedin — APPROVED** [review]

**Pull request merged**